### PR TITLE
65292_user_comment_on_news

### PR DESCRIPTION
**Pull request by yasin-buet**
cc1

### CODE REVIEW

**Comment by asifbuetcse** [comment]
OK

File: app/src/Controller/CommentsController.php [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+ * @property \App\Model\Table\CommentsTable $Comments
+ */
+class CommentsController extends AppController
+{
+    public function isAuthorized($user)
+    {
+        if (in_array($this->request->action, ['index', 'view', 'add'])) {
+            return true;
+        }
+        if (($this->request->session()->read('Auth.User.id') == $user['id']) &&
```
Comment (tuandv2611) [inline app/src/Controller/CommentsController.php]: blank line

File: app/src/Controller/CommentsController.php [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+            (in_array($this->request->action, ['edit', 'delete']))) {
```
Comment (tuandv2611) [inline app/src/Controller/CommentsController.php]: && in_array($this->request->action, ['edit', 'delete'])) {

File: app/src/Model/Table/CommentsTable.php [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+     * @return void
+     */
+    public function initialize(array $config)
+    {
+        parent::initialize($config);
+
+        $this->table('comments');
+        $this->displayField('id');
+        $this->primaryKey('id');
+
```
Comment (tuandv2611) [inline app/src/Model/Table/CommentsTable.php]: remove blank line

File: app/src/Model/Table/CommentsTable.php [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        $this->addBehavior('Timestamp');
+
```
Comment (tuandv2611) [inline app/src/Model/Table/CommentsTable.php]: remove blank line

File: app/src/Template/Admin/Comments/index.ctp [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        <tbody>
+            <?php foreach ($comments as $comment): ?>
+            <tr>
+                <td><?= $this->Number->format($comment->id) ?></td>
+                <td><?= h($comment->body) ?></td>
+                <td>
+                    <?= $comment->has('user') ? $this->Html->link($comment->user->id, [
+                        'controller' => 'Users',
+                        'action' => 'view', $comment->user->id
+                        ]) : '' ?>
```
Comment (tuandv2611) [inline app/src/Template/Admin/Comments/index.ctp]: indent

File: app/src/Template/News/view.ctp [hunk 6]
```diff
@@ -56,3 +57,13 @@
...
     </div>
 </div>
+<div class="btn-group col-sm-12">
+        <?= $this->Html->link('Add A Comment', [
+            'controller' => 'Comments',
+            'action' => 'add',
+            $news->id,
+        ], [
+            'class' => 'btn btn-primary col-sm-5',
+            'role' => 'button'
```
Comment (tuandv2611) [inline app/src/Template/News/view.ctp]: take "," into last element of multi Array `'role' => 'button',`

File: app/src/Template/News/view.ctp [hunk 7]
```diff
@@ -40,14 +39,16 @@
                 <td><?= h($comments->id) ?></td>
                 <td><?= h($comments->body) ?></td>
                 <td><?= h($comments->user_id) ?></td>
-                <td><?= h($comments->news_id) ?></td>
                 <td><?= h($comments->created) ?></td>
                 <td><?= h($comments->modified) ?></td>
                 <td class="actions">
                     <?= $this->Html->link(__('View'), ['controller' => 'Comments', 'action' => 'view', $comments->id]) ?>
-                    <?php if ($this->request->session()->read('Auth.User.role') == 'admin'): ?>
+                    <?php if (($this->request->session()->read('Auth.User.role') == 'admin') || ($this->request->session()->read('Auth.User.id') == $comments->user_id)): ?>
```
Comment (tuandv2611) [inline app/src/Template/News/view.ctp]: Too long, format again

File: app/src/Template/Admin/News/view.ctp [hunk 8]
```diff
@@ -47,7 +47,10 @@
                     <?= $this->Html->link(__('View'), ['controller' => 'Comments', 'action' => 'view', $comments->id]) ?>
                     <?php if ($this->request->session()->read('Auth.User.role') == 'admin'): ?>
                         <?= $this->Html->link(__('Edit'), ['controller' => 'Comments', 'action' => 'edit', $comments->id]) ?>
-                        <?= $this->Form->postLink(__('Delete'), ['controller' => 'Comments', 'action' => 'delete', $comments->id], ['confirm' => __('Are you sure you want to delete # {0}?', $comments->id)]) ?>
+                        <?= $this->Form->postLink(__('Delete'),
+                        ['action' => 'delete', $comments->id],
```
Comment (tuandv2611) [inline app/src/Template/Admin/News/view.ctp]: indent

File: app/src/Template/Admin/Comments/index.ctp [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+                </td>
+                <td><?= h($comment->created) ?></td>
+                <td><?= h($comment->modified) ?></td>
+                <td class="actions">
+                    <?= $this->Html->link(__('View'), ['action' => 'view', $comment->id]) ?>
+                    <?= $this->Html->link(__('Edit'), ['action' => 'edit', $comment->id]) ?>
+                    <?= $this->Form->postLink(__('Delete'),
+                        ['action' => 'delete', $comment->id],
+                        ['confirm' => __('Are you sure you want to delete # {0}?', $comment->id)])
+                    ?>
```
Comment (tuandv2611) [inline app/src/Template/Admin/Comments/index.ctp]: break `)  => ) ?>`

File: app/src/Controller/CommentsController.php [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+            if ($this->Comments->save($comment)) {
+                $this->Flash->success(__('The comment has been saved.'));
+                return $this->redirect(['controller' => 'News', 'action' => 'index']);
+            } else {
+                $this->Flash->error(__('The comment could not be saved. Please, try again.'));
+            }
+        }
+
+        $users = $this->Comments->Users->find('list', ['limit' => 200]);
+        $news = $this->Comments->News->find('list', ['limit' => 200]);
```
Comment (hieubm) [inline app/src/Controller/CommentsController.php]: do you need list of user and news here ?

File: app/src/Template/Admin/Comments/edit.ctp [hunk 11]
```diff
@@ -0,0 +1,12 @@
+<div class="comments form large-9 medium-8 columns content">
+    <?= $this->Form->create($comment) ?>
+    <fieldset>
+        <legend><?= __('Edit Comment') ?></legend>
+        <?php
+            echo $this->Form->input('body', ['class' => 'form-control']);
+            echo $this->Form->input('news_id', ['options' => $news, 'class' => 'form-control']);
```
Comment (hieubm) [inline app/src/Template/Admin/Comments/edit.ctp]: what do you think if combine add/edit to 1 form.ctp?
btw, admin shouldn't change new_id of comment in this edit form,

File: app/src/Template/Admin/Comments/index.ctp [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+                <th class="actions"><?= __('Actions') ?></th>
+            </tr>
+        </thead>
+        <tbody>
+            <?php foreach ($comments as $comment): ?>
+            <tr>
+                <td><?= $this->Number->format($comment->id) ?></td>
+                <td><?= h($comment->body) ?></td>
+                <td>
+                    <?= $comment->has('user') ? $this->Html->link($comment->user->id, [
```
Comment (hieubm) [inline app/src/Template/Admin/Comments/index.ctp]: any comment doesn't have user ?

File: app/src/Template/Admin/News/view.ctp [hunk 13]
```diff
@@ -56,3 +59,14 @@
         <?php endif; ?>
     </div>
 </div>
+<div class="btn-group col-sm-12">
+        <?= $this->Html->link('Add A Comment', [
```
Comment (hieubm) [inline app/src/Template/Admin/News/view.ctp]: The indent is too many

File: app/src/Controller/CommentsController.php [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+ */
+class CommentsController extends AppController
+{
+    public function isAuthorized($user)
+    {
+        if (in_array($this->request->action, ['index', 'view', 'add'])) {
+            return true;
+        }
+
+        if (($this->request->session()->read('Auth.User.id') == $user['id'])
```
Comment (hieubm) [inline app/src/Controller/CommentsController.php]: Can you try
`$this->AuthUser->id();`

File: app/src/Template/News/view.ctp [hunk 15]
```diff
@@ -40,14 +39,17 @@
                 <td><?= h($comments->id) ?></td>
                 <td><?= h($comments->body) ?></td>
                 <td><?= h($comments->user_id) ?></td>
-                <td><?= h($comments->news_id) ?></td>
                 <td><?= h($comments->created) ?></td>
                 <td><?= h($comments->modified) ?></td>
                 <td class="actions">
                     <?= $this->Html->link(__('View'), ['controller' => 'Comments', 'action' => 'view', $comments->id]) ?>
-                    <?php if ($this->request->session()->read('Auth.User.role') == 'admin'): ?>
+                    <?php if (($this->request->session()->read('Auth.User.role') == 'admin')
```
Comment (hieubm) [inline app/src/Template/News/view.ctp]: can you try
$this->AuthUser->user('role')

File: app/src/Controller/CommentsController.php [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+    public function add($newsId)
+    {
+        $comment = $this->Comments->newEntity();
+        if ($this->request->is('post')) {
+            $comment['user_id'] = $this->Auth->user('id');
+            $comment['news_id'] = $newsId;
+            $comment = $this->Comments->patchEntity($comment, $this->request->data);
+            if ($this->Comments->save($comment)) {
+                $this->Flash->success(__('The comment has been saved.'));
+                return $this->redirect(['controller' => 'News', 'action' => 'index']);
```
Comment (hieubm) [inline app/src/Controller/CommentsController.php]: it's better to redirect back to the news which is commented

File: app/src/Controller/CommentsController.php (hunk 10 above)
Comment (yasin-buet) [inline app/src/Controller/CommentsController.php]: @hieubm 
removed

File: app/src/Template/News/view.ctp (hunk 15 above)
Comment (yasin-buet) [inline app/src/Template/News/view.ctp]: @hieubm 
In view auth-helper can not be used. 
That's only for controllers.

Comment (hieubm) [inline app/src/Template/News/view.ctp]: Can you try 
`$this->set('authUser', $this->Auth->user());`
in AppController, maybe in beforeFilter
and, in view, you can use: $this->authUser->xxx

File: app/src/Controller/Admin/CommentsController.php [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+    public function add($newsId)
+    {
+        $comment = $this->Comments->newEntity();
+        if ($this->request->is('post')) {
+            $comment['user_id'] = $this->Auth->user('id');
+            $comment['news_id'] = $newsId;
+            $comment = $this->Comments->patchEntity($comment, $this->request->data);
+            if ($this->Comments->save($comment)) {
+                $this->Flash->success(__('The comment has been saved.'));
+                return $this->redirect(['controller' => 'News', 'action' => 'index']);
```
Comment (hieubm) [inline app/src/Controller/Admin/CommentsController.php]: It should rediect to $newsId page, not new index

File: app/src/Controller/Admin/CommentsController.php [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+            if ($this->Comments->save($comment)) {
+                $this->Flash->success(__('The comment has been saved.'));
+                return $this->redirect(['action' => 'index']);
+            } else {
+                $this->Flash->error(__('The comment could not be saved. Please, try again.'));
+            }
+        }
+
+        $users = $this->Comments->Users->find('list', ['limit' => 200]);
+        $news = $this->Comments->News->find('list', ['limit' => 200]);
```
Comment (hieubm) [inline app/src/Controller/Admin/CommentsController.php]: Do you need these lists here?
When I commented about a problem, you should check all other parts of your code to fix all related problems

File: app/src/Controller/CommentsController.php [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+ */
+class CommentsController extends AppController
+{
+    public function isAuthorized($user)
+    {
+        if (in_array($this->request->action, ['index', 'view', 'add'])) {
+            return true;
+        }
+
+        if (($this->Auth->user('id') == $user['id'])
```
Comment (hieubm) [inline app/src/Controller/CommentsController.php]: the $user['id'] is current user id,
and $this->Auth->user('id') is current user id too
it always true

Please check this case:
user 1: create a comment id = 1
user 2, access: comments/edit/1, and he also can edit it,
please fix it

File: app/src/Controller/CommentsController.php (hunk 10 above)
Comment (hieubm) [inline app/src/Controller/CommentsController.php]: I still see it here

File: app/src/Template/Comments/edit.ctp [hunk 20]
```diff
@@ -0,0 +1,12 @@
+<div class="comments form large-9 medium-8 columns content">
+    <?= $this->Form->create($comment) ?>
+    <fieldset>
+        <legend><?= __('Edit Comment') ?></legend>
+        <?php
+            echo $this->Form->input('body', ['class' => 'form-control']);
+            echo $this->Form->input('news_id', ['options' => $news, 'class' => 'form-control']);
```
Comment (hieubm) [inline app/src/Template/Comments/edit.ctp]: why news_id is still here?
and edit and add should be combined to form.ctp

File: app/src/Controller/AppController.php [hunk 21]
```diff
@@ -27,7 +27,7 @@
  */
 class AppController extends Controller
 {
-
+    var $helpers = array('Form', 'Html');
```
Comment (hieubm) [inline app/src/Controller/AppController.php]: it's better to have an enter after this

File: app/src/Template/Admin/Comments/index.ctp [hunk 22]
```diff
@@ -0,0 +1,55 @@
...
+                <td><?= h($comment->body) ?></td>
+                <td>
+                    <?= $this->Html->link($comment->user->id, [
+                        'controller' => 'Users',
+                        'action' => 'view', $comment->user->id
+                        ]) : ''
+                    ?>
+                </td>
+                <td>
+                    <?= $comment->has('news') ? $this->Html->link($comment->news->title,
```
Comment (hieubm) [inline app/src/Template/Admin/Comments/index.ctp]: Does any comment not belong to any news ?
I think $comment->has('news') is not neccessry

File: app/src/Template/Admin/Comments/view.ctp [hunk 23]
```diff
@@ -0,0 +1,39 @@
...
+                <?= $comment->has('user') ? $this->Html->link($comment->user->id, [
+                    'controller' => 'Users',
+                    'action' => 'view', $comment->user->id
+                ]) : '' ?>
+            </td>
+        </tr>
+        <tr>
+            <th><?= __('News') ?></th>
+            <td>
+                <?= $comment->has('news') ? $this->Html->link($comment->news->title, [
```
Comment (hieubm) [inline app/src/Template/Admin/Comments/view.ctp]: don't forget to fix these things (facepalm)

File: app/src/Template/News/view.ctp [hunk 24]
```diff
@@ -56,3 +58,13 @@
         <?php endif; ?>
     </div>
 </div>
+<div class="btn-group col-sm-12">
+        <?= $this->Html->link('Add A Comment', [
```
Comment (hieubm) [inline app/src/Template/News/view.ctp]: this one has too many indent too
please self-check your pull request carefully (bow)

File: app/src/Template/Admin/Comments/view.ctp [hunk 25]
```diff
@@ -0,0 +1,39 @@
...
+            <th><?= __('Body') ?></th>
+            <td><?= h($comment->body) ?></td>
+        </tr>
+        <tr>
+            <th><?= __('User') ?></th>
+            <td>
+                <?= $this->Html->link($comment->user->id, [
+                    'controller' => 'Users',
+                    'action' => 'view', $comment->user->id
+                ]) : '' ?>
```
Comment (hieubm) [inline app/src/Template/Admin/Comments/view.ctp]: ` : '' `
does it run ?

File: app/src/Template/Admin/Comments/index.ctp [hunk 26]
```diff
@@ -0,0 +1,55 @@
...
+                        'controller' => 'Users',
+                        'action' => 'view', $comment->user->id
+                        ]) : ''
+                    ?>
+                </td>
+                <td>
+                    <?= $this->Html->link($comment->news->title,
+                        ['controller' => 'News',
+                        'action' => 'view', $comment->news->id
+                    ]) : '' ?>
```
Comment (hieubm) [inline app/src/Template/Admin/Comments/index.ctp]: why don't you fix this one and the above one
(facepalm)

File: app/src/Template/Comments/form.ctp [hunk 27]
```diff
@@ -0,0 +1,11 @@
+<div class="comments form large-9 medium-8 columns content">
+    <?= $this->Form->create($comment) ?>
+    <fieldset>
+        <legend><?= __('Comment') ?></legend>
+        <?php
```
Comment (hieubm) [inline app/src/Template/Comments/form.ctp]: what about <?= ...

**Pull request merged**